### ISSUE TITLE
Question about activation function choice:

### ISSUE THREAD

**Issue by ElliottDyson** [state-spaces/mamba#157]
Just a point of curiosity really. I noticed in your code that either ReLU or Swish is used. I understand the choice of including ReLU as it is commonly accepted as being very efficient. However the usage of Swish I am a little more perplexed by when SwishGLU or Mish have previously shown better performance on transformer architectures? 

I'd be very curious to see how this model performs if trained with the activation function set as Mish given it seems to be the best performing activation function for transformer encoder feed forward networks (as far as I'm aware).

**Comment by ElliottDyson** [comment]
> Good idea! We haven't experimented with it. To be honest I think these differences tend to get washed out with scale, but maybe not.

I understand that the new architecture you have developed is different from what I am used to. I started learning with transformers rather than CNNs, which is quite unusual (I know 😅). I am still trying to fully understand how this new architecture works. However, I found that when I moved from ReLU to Mish in the feed-forward network of my transformer encoder, I had significant improvements in the validation loss landscape. This was with a 16 attention head, 512 embed dim, 10 layer deep model, which is not that insignificant in scale.

The reason I mentioned that I am still wrapping my head around this new architecture is that I am not quite sure about the intricacies of how the activation function interacts with the rest of the network. With a transformer encoder, I understand why a good choice of activation function somewhat scales with model size. It may just be insignificant at scale with this architecture.

If you decide to run the test, I would appreciate an @ to see the results. Unfortunately, I do not have the compute available for such a task, else I would run it myself out of curiosity.

I also wanted to thank you for the amazing work and research you have done here to develop this new architecture. I am looking forward to understanding it more thoroughly and hopefully being able to contribute to its further development in any way possible. This is really looking like it's going to be the next big architecture, like transformers were to LSTMs.

**Comment by Maykeye** [comment]
> Good idea! We haven't experimented with it. To be honest I think these differences tend to get washed out with scale, but maybe not.

[Paper in point](https://arxiv.org/pdf/2310.04564.pdf). Switching activation function back to ReLU was one of the method to speed up [llama](https://huggingface.co/SparseLLM/ReluLLaMA-7B). They got worse results on benchmarks, but they finetuned model on 5B tokens, when Llama was trained on trillions, so 🤷

Helps that most of activation functions are "Look like ReLU, but differentiable and slightly different around zero where most values are". 

> SwishGLU

It also requires parameters so it's not surprised that it will be better. 

Though it actually interesting if these activation functions are good places to inject some sort of adapters.

**Comment by ElliottDyson** [comment]
> > Good idea! We haven't experimented with it. To be honest I think these differences tend to get washed out with scale, but maybe not.
> 
> [Paper in point](https://arxiv.org/pdf/2310.04564.pdf). Switching activation function back to ReLU was one of the method to speed up [llama](https://huggingface.co/SparseLLM/ReluLLaMA-7B). They got worse results on benchmarks, but they finetuned model on 5B tokens, when Llama was trained on trillions, so 🤷
> 
> Helps that most of activation functions are "Look like ReLU, but differentiable and slightly different around zero where most values are". 
> 
> > SwishGLU
> 
> It also requires parameters so it's not surprised that it will be better. 
> 
> Though it actually interesting if these activation functions are good places to inject some sort of adapters.
> 

What is the meaning behind them being good places to inject adapters?

Also, I communicated with someone that's currently doing their research on activation functions and they were able to provide me one that was working significantly better than Mish, well, with a transformer network on the tiny stories dataset anyways, not that the specific dataset should matter much. But the equation is: tanh(X1) * (X2)^2 in case you plan on testing different activation functions on the network.

**Comment by Maykeye** [comment]
> What is the meaning behind them being good places to inject adapters?

Long story: arxiv:1902.00751. 
Short story:  if LoRA replaces  `XW` with  `XW  +  XAB` and sees itself more as replacement for matrix, adapters replace any block `f(X)` with `f(X) + g(XA)B`  (so they apply non-linearity after down proj, unlike lora, but in both cases original model is frozen and you train only small number of new parameters).

So  I was wondering if its' possible to easily replace SiLU with something like `SiLU(x) + trainable(SwiGLU(x))` and see how it perform after the training.
Well, apparently the code really wants activation function to be silu in many places. And if fast path and special [causal_conv_1d](https://github.com/state-spaces/mamba/blob/009bec5ee37f586844a3fc89c040a9c1a9d8badf/mamba_ssm/modules/mamba_simple.py#L168) can be disabled, SiLU also seems to be baked into cuda kernel  [selective scan](https://github.com/state-spaces/mamba/blob/009bec5ee37f586844a3fc89c040a9c1a9d8badf/mamba_ssm/modules/mamba_simple.py#L188)
